cmd/kube-apiserver/app/server.go

```diff
@@ -145,6 +145,25 @@ func Run(s *options.APIServer) error {
 		glog.Fatalf("Failed to start kubelet client: %v", err)
 	}
 
+	if s.StorageConfig.DeserializationCacheSize == 0 {
+		// When size of cache is not explicitly set, estimate its size based on
+		// target memory usage.
+		glog.V(2).Infof("Initalizing deserialization cache size based on %dMB limit", s.TargetRAMMB)
+
+		// This is the heuristics that from memory capacity is trying to infer
+		// the maximum number of nodes in the cluster and set cache sizes based
+		// on that value.
+		// From our documentation, we officially recomment 120GB machines for
+		// 2000 nodes, and we scale from that point. Thus we assume ~60MB of
+		// capacity per node.
```

Comment:
I also considered it (when thinking about it for the sizes of caches in "Cacher"), but decided that it's not the best option for now. First of all, because the target memory is not super accurate now (it actually is more-or-less target memory for all components on master machine). Second, because it will result in waaaaay bigger memory usage for larger clusters, which is also not very good in my opinion.

So I would prefer to leave it as is (especially since we want to cherrypick) and probably solve it better later (for 1.5 maybe?). [Though I extended the TODO so that it contains what you basically wrote].